src/seq_analysis/orf.rs

```diff
@@ -0,0 +1,163 @@
+                   min_len: usize)
+                   -> Self {
+        Finder {
+            start_codons: start_codons.into_iter()                          // Convert start_ and
+                                      .map(|x| {                            // stop_codons from 
+                                          x.into_iter()                     // Vec<&[u8;3]> to
+                                           .map(|&x| x as u8)               // Vec<VecDeque<u8>>
+                                           .collect::<VecDeque<u8>>()       // so they can be
+                                      })                                    // easily compared
+                                      .collect(), // with codon built
+            stop_codons: stop_codons.into_iter()                            // from IntoTextIterator
+                                    .map(|x| {                              // object.
+                                        x.into_iter()
+                                         .map(|&x| x as u8)
+                                         .collect::<VecDeque<u8>>()
+                                    })
+                                    .collect(),
+            min_len: min_len,
+        }
+    }
+
+    /// Find all orfs in the given sequence
+    pub fn find_all<'a, I: IntoTextIterator<'a>>(&'a self, seq: I) -> Matches<I::IntoIter> {
+        Matches {
+            finder: self,
+            state: State::new(),
+            seq: seq.into_iter().enumerate(),
+        }
+    }
+}
+
+/// An orf representation with start and end position of said orf,
+/// as well as offset of the reading frame (1,2,3) and strand location 
+// (current: +, reverse complementary: -).
+pub struct Orf {
+    pub start: usize,
+    pub end: usize,
+    pub offset: i8,
+}
+
+/// The current algorithm state.
+struct State {
+    start_pos: [Option<usize>; 3],
+    in_orf: [bool; 3],
+    codon: VecDeque<u8>,
+}
+
+
+impl State {
+    /// Create new state.
+    pub fn new() -> Self {
+        State {
+            start_pos: [None, None, None],
+            in_orf: [false, false, false],
+            codon: VecDeque::new(),
+        }
+    }
+}
+
+
+/// Iterator over offset, start position, end position and sequence of matched orfs.
+pub struct Matches<'a, I: TextIterator<'a>> {
+    finder: &'a Finder,
+    state: State,
+    seq: iter::Enumerate<I>,
+}
+
+impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+    type Item = Orf;
+
+    fn next(&mut self) -> Option<Orf> {
+
+        let mut result: Option<Orf> = None;
+        let mut offset: usize;
+
+        for (index, &nuc) in self.seq.by_ref() {
+
+            // update the codon
+            if self.state.codon.len() >= 3 {
+                self.state.codon.pop_front();
+            }
+            self.state.codon.push_back(nuc);
+            offset = (index + 1) % 3;
+
+            // inside orf
+            if self.state.in_orf[offset] {
```

Comment:
Since you have now options in start_pos, in_orf is not needed anymore. You can instead check if the option is none.